dbt/adapters/cache.py

```diff
@@ -0,0 +1,463 @@
+    """
+    def __init__(self, schema, identifier, inner):
+        self.schema = schema
+        self.identifier = identifier
+        self.referenced_by = {}
+        self.inner = inner
+
+    def __str__(self):
+        return (
+            'CachedRelation(schema={}, identifier={}, inner={})'
+        ).format(self.schema, self.identifier, self.inner)
+
+    def __copy__(self):
+        new = self.__class__(self.schema, self.identifier)
+        new.__dict__.update(self.__dict__)
+        return new
+
+    def __deepcopy__(self, memo):
+        new = self.__class__(self.schema, self.identifier)
+        new.__dict__.update(self.__dict__)
+        new.referenced_by = deepcopy(self.referenced_by, memo)
+        new.inner = self.inner.incorporate()
+
+    def is_referenced_by(self, key):
+        return key in self.referenced_by
+
+    def key(self):
+        """Get the ReferenceKey that represents this relation
+
+        :return ReferenceKey: A key for this relation.
+        """
+        return ReferenceKey(self.schema, self.identifier)
+
+    def add_reference(self, referrer):
+        """Add a reference from referrer to self, indicating that if this node
+        were drop...cascaded, the referrer would be dropped as well.
+
+        :param CachedRelation referrer: The node that refers to this node.
+        """
+        self.referenced_by[referrer.key()] = referrer
+
+    def collect_consequences(self):
+        """Recursively collect a set of ReferenceKeys that would
+        consequentially get dropped if this were dropped via
+        "drop ... cascade".
+
+        :return Set[ReferenceKey]: All the relations that would be dropped
+        """
+        consequences = {self.key()}
+        for relation in self.referenced_by.values():
+            consequences.update(relation.collect_consequences())
+        return consequences
+
+    def release_references(self, keys):
+        """Non-recursively indicate that an iterable of ReferenceKey no longer
+        exist. Unknown keys are ignored.
+
+        :param Iterable[ReferenceKey] keys: The keys to drop.
+        """
+        keys = set(self.referenced_by) & set(keys)
+        for key in keys:
+            self.referenced_by.pop(key)
+
+    def rename(self, new_relation):
+        """Rename this cached relation to new_relation, updating the inner
+        Reference as well.
+        Note that this will change the output of key(), all refs must be
+        updated!
+
+        :param ReferenceKey new_relation: The new name to apply to the relation
+        """
+        self.schema = new_relation.schema
+        self.identifier = new_relation.identifier
+        # rename our inner value as well
+        if self.inner:
+            # Relations store this stuff inside their `path` dict. But they
+            # also store a table_name, and usually use it in their  .render(),
+            # so we need to update that as well. It doesn't appear that
```

Comment:
`table_name` used to conditionally be `name + '__dbt_tmp`, but we've since removed that, so now `table_name == identifier` I believe. One thing to watch out for here might be ephemeral models... I need to do more digging to see if they're a relevant concern here, but wanted to surface it.